### ISSUE TITLE
REFL037 should not warn here 

### ISSUE THREAD

**Issue by JohanLarsson** [DotNetAnalyzers/ReflectionAnalyzers#161]
```cs
typeof(BinaryExpression).Assembly.GetType("System.Linq.Expressions.AssignBinaryExpression", throwOnError: true)
```

Invisible strikes again?

**Comment by JohanLarsson** [comment]
```cs
typeof(AdornedElementPlaceholder).Assembly.GetType("MS.Internal.Controls.TemplatedAdorner", throwOnError: true);
```

**Comment by brian-reichle** [comment]
Maybe the only thing you can assert is that the 'type name' is well formed :)

**Comment by JohanLarsson** [comment]
Yes we need to disable it for now, it would be a useful analyzer so nice if we can find a way to make it work.
Maybe we can re-enable it when we upgrade to Roslyn 3.x and we can detect if internals are visible.

I've been thinking about a huge hack where we pack a small exe in the analyzer dll and use the exe to dump the internals using standard reflection.

**Comment by jnm2** [comment]
The analyzer needs to be able to quantify what it doesn't know. Right now it's based on whether it can prove the member exists, but it should be based on whether it can prove the member *doesn't* exist.

Scenarios it needs to detect to properly quantify the level of uncertainty:

- Public members are always reported by Roslyn. ✔ Can prove nonexistence of public members.
- Roslyn semantic model is reporting private members? ✔ Can prove nonexistence of private members.
- Otherwise, it's not project reference. Must be an assembly reference:
  - Assembly cannot be resolved? ❌ Cannot prove nonexistence of private members.
  - Or, resolved assembly is a reference assembly? ❌ Cannot prove nonexistence of private members.
  - Or, resolved assembly does not end up in build output (it will likely run against a different framework version with a different set of private members)? ❌ Cannot prove nonexistence of private members.
  - Read the metadata in the resolved assembly. ✔ Can prove nonexistence of private members.

REFL037 should only be reported when you reach a green checkmark and prove that the member does not exist.

**Comment by jnm2** [comment]
<code>(<a href="http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/CompilationOptions.cs,3de3a8f794790b63,references">compilation.Options.MetadataImportOptions</a> & MetadataImportOptions.Internal) != 0</code> tells you whether you can stop at the second green checkmark at the top.

**Comment by jnm2** [comment]
Wait, it's not a `[Flags]` enum.

[`compilation.Options.MetadataImportOptions`](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/CompilationOptions.cs,3de3a8f794790b63,references) tells you whether you can stop at the second green checkmark at the top, but it also has three levels, not two.

**Comment by jnm2** [comment]
I'd expect the analyzer to either skip that set then, or else only skip that step if you can't find and invoke the members via reflection.

**Comment by JohanLarsson** [comment]
Yes, I'll disable this analyzer for now until we have better handling. As it is now it is 100% annoying.

**Comment by jnm2** [comment]
I updated the [logic](https://github.com/DotNetAnalyzers/ReflectionAnalyzers/issues/161#issuecomment-432255103) that I'd use.

**Comment by JohanLarsson** [comment]
I had a horrible bug, closed this issue with the fix for that. We should resurrect the discussion as it is still valid.
